core/src/main/java/org/elasticsearch/search/SearchModule.java

```diff
@@ -432,23 +432,73 @@ public AggregationSpec addResultReader(String writeableName, Writeable.Reader<?
 
     /**
      * Register a pipeline aggregation.
-     *
-     * @param reader reads the aggregation builder from a stream
-     * @param internalReader reads the {@link PipelineAggregator} from a stream
-     * @param bucketReader reads the {@link InternalAggregation} that represents a bucket in this aggregation from a stream
-     * @param aggregationParser reads the aggregation builder from XContent
-     * @param aggregationName names by which the aggregation may be parsed. The first name is special because it is the name that the reader
-     *        is registered under.
      */
-    public void registerPipelineAggregation(Writeable.Reader<? extends PipelineAggregationBuilder> reader,
-            Writeable.Reader<? extends PipelineAggregator> internalReader, Writeable.Reader<? extends InternalAggregation> bucketReader,
-            PipelineAggregator.Parser aggregationParser, ParseField aggregationName) {
+    public void registerPipelineAggregation(PipelineAggregationSpec spec) {
         if (false == transportClient) {
-            namedWriteableRegistry.register(PipelineAggregationBuilder.class, aggregationName.getPreferredName(), reader);
-            pipelineAggregationParserRegistry.register(aggregationParser, aggregationName);
+            namedWriteableRegistry.register(PipelineAggregationBuilder.class, spec.name.getPreferredName(), spec.builderReader);
+            pipelineAggregationParserRegistry.register(spec.parser, spec.name);
+        }
+        for (Map.Entry<String, Writeable.Reader<? extends PipelineAggregator>> resultReader : spec.resultReaders.entrySet()) {
+            namedWriteableRegistry.register(PipelineAggregator.class, resultReader.getKey(), resultReader.getValue());
+        }
+        for (Map.Entry<String, Writeable.Reader<? extends InternalAggregation>> bucketReaders : spec.bucketReaders.entrySet()) {
+            namedWriteableRegistry.register(InternalAggregation.class, bucketReaders.getKey(), bucketReaders.getValue());
+        }
+    }
+
+    public static class PipelineAggregationSpec {
```

Comment:
I'm a bit torn on this. I like the similarity to the other Aggregations and this does also leave it open for pipeline aggregators to have multiple result readers if they need to, but on the other hand at the moment it's not currently needed. I'd say I'm slightly leaning towards keeping it like this but I am happy with either way so coders choice. 😄